### PR TITLE
[SPARK-41180][SQL] Reuse `INVALID_SCHEMA` instead of `_LEGACY_ERROR_TEMP_1227`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims rename `_LEGACY_ERROR_TEMP_1227` to `INVALID_SCHEMA`

### Why are the changes needed?
Proper names of error classes to improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

**Comment by MaxGekk** [comment]
@LuciferYang Could you resolve conflicts, please.

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -641,6 +641,12 @@
     ],
     "sqlState" : "42000"
   },
+  "INVALID_DATA_TYPE_SCHEMA" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:644]: We already have the error class `INVALID_SCHEMA`. Can you re-use it or make it more generic? cc @srielau

File: sql/core/src/test/resources/sql-tests/results/csv-functions.sql.out [hunk 2]
```diff
@@ -43,7 +43,7 @@ struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
 {
-  "errorClass" : "_LEGACY_ERROR_TEMP_1227",
+  "errorClass" : "INVALID_DATA_TYPE_SCHEMA",
   "messageParameters" : {
     "e1" : "\n[PARSE_SYNTAX_ERROR] Syntax error at or near 'InvalidType': extra input 'InvalidType'(line 1, pos 2)\n\n== SQL ==\na InvalidType\n--^^^\n",
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/csv-functions.sql.out:48]: How about to unwrap this error? When both attempts fail, just bypass the first one to users.

**Comment by LuciferYang** [comment]
two test case failed, will fix later

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -780,8 +780,21 @@
...
+      "The input schema <inputSchema> is not a valid schema string."
+    ],
+    "subClass" : {
+      "PARSE_ERROR" : {
+        "message" : [
+          "Cannot parse the schema: ",
+          "<reason>"
+        ]
+      },
+      "UNEXPECTED_INPUT_TYPE" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:792]: How about `NON_STRING_LITERAL`?

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -780,8 +780,21 @@
...
   "INVALID_SCHEMA" : {
     "message" : [
-      "The expression <expr> is not a valid schema string."
-    ]
+      "The input schema <inputSchema> is not a valid schema string."
+    ],
+    "subClass" : {
+      "PARSE_ERROR" : {
+        "message" : [
+          "Cannot parse the schema: ",
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:788]:
```suggestion
          "Cannot parse the schema:",
```

Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json]: Thanks

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json:792]: renamed

Comment (LuciferYang) [reply core/src/main/resources/error/error-classes.json]: [af157d2](https://github.com/apache/spark/pull/38754/commits/af157d2bf34489ebbd116f6077eb13c5287971cc) rename ` INVALID_SCHEMA.UNEXPECTED_INPUT_TYPE` to `INVALID_SCHEMA.NON_STRING_LITERAL`

File: sql/core/src/test/resources/sql-tests/results/csv-functions.sql.out [hunk 5]
```diff
@@ -43,11 +43,10 @@ struct<>
...
 org.apache.spark.sql.AnalysisException
 {
-  "errorClass" : "_LEGACY_ERROR_TEMP_1227",
+  "errorClass" : "INVALID_SCHEMA.PARSE_ERROR",
   "messageParameters" : {
-    "e1" : "\n[PARSE_SYNTAX_ERROR] Syntax error at or near 'InvalidType': extra input 'InvalidType'(line 1, pos 2)\n\n== SQL ==\na InvalidType\n--^^^\n",
-    "e2" : "\nDataType invalidtype is not supported.(line 1, pos 2)\n\n== SQL ==\na InvalidType\n--^^^\n",
-    "msg" : "Cannot parse the data type: "
+    "inputSchema" : "\"a InvalidType\"",
+    "reason" : "\n[PARSE_SYNTAX_ERROR] Syntax error at or near 'InvalidType': extra input 'InvalidType'(line 1, pos 2)\n\n== SQL ==\na InvalidType\n--^^^\n"
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/csv-functions.sql.out:49]: Let's bypass `SparkThrowable` to users. I don't think the wrapping is useful.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 6]
```diff
@@ -160,8 +157,8 @@ object DataType {
         try {
           fallbackParser(schema)
         } catch {
-          case NonFatal(e2) =>
-            throw QueryCompilationErrors.failedFallbackParsingError(errorMsg, e1, e2)
+          case NonFatal(_) =>
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:160]: Do you mean if `e2` is `NonFatal ` and `e1` is  `SparkThrowable`, then re-throw it,  and only call `QueryCompilationErrors.schemaFailToParseError` when e1 is not `SparkThrowable`?

[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:160]: Yep

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 7]
```diff
@@ -443,19 +443,31 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
         start = 0,
         stop = 18
       )
     )
 
-    val errMsg2 = intercept[AnalysisException] {
-      df3.selectExpr("""from_json(value, 'time InvalidType')""")
-    }
-    assert(errMsg2.getMessage.contains("DataType invalidtype is not supported"))
+    checkError(
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:455]: The case corresponding to `e` is `SparkThrowable`

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 8]
```diff
@@ -991,22 +1003,47 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
   test("SPARK-33286: from_json - combined error messages") {
     val df = Seq("""{"a":1}""").toDF("json")
     val invalidJsonSchema = """{"fields": [{"a":123}], "type": "struct"}"""
-    val errMsg1 = intercept[AnalysisException] {
-      df.select(from_json($"json", invalidJsonSchema, Map.empty[String, String])).collect()
-    }.getMessage
-    assert(errMsg1.contains("""Failed to convert the JSON string '{"a":123}' to a field"""))
+    val invalidJsonSchemaReason = "Failed to convert the JSON string '{\"a\":123}' to a field."
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:1006]: The following two case corresponding to `e` is  not SparkThrowable

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @LuciferYang.

**Comment by LuciferYang** [comment]
Thanks @MaxGekk

**Pull request closed without merging**